blogs/ulysses-offload/README.md

```diff
@@ -0,0 +1,263 @@
+},
+
+journal={https://arxiv.org/abs/2408.16978},
+
+year={2024}
+
+}
+
+## Introduction
+
+In the rapidly evolving field of generative AI and scientific ML, the
+ability to train large (language) models with ultra-long context
+capabilities is becoming increasingly important. These models are
+essential for a variety of complex tasks, ranging from understanding
+lengthy documents to image and video generation to processing extensive
+sequences in computational biology. However, training such models
+efficiently poses significant challenges due to the enormous GPU
+resources and memory required.
+
+Building on our previous project, DeepSpeed Ulysses, which focuses on
+system optimizations for training extremely long sequence transformer
+models, we are excited to present the Fully Pipelined Distributed
+Transformer (FPDT), also known as Ulysses-Offload, in this release. FPDT
+is an innovative, resource-efficient technique that offers comparable
+benefits to DeepSpeed Ulysses and other previous long-context
+optimization methods, but with a modest hardware budget. FPDT makes
+ultra long-context large language models (LLM) training and finetuning
+accessible to everyone, regardless of GPU resources. FPDT enables
+training with context lengths of up to 2 million tokens using just 4
+NVIDIA A100-40GB GPUs. FPDT supports 16x longer sequence lengths at 55% 
+Model FLOPs Utilization (MFU) than NVIDIA Megatron-SP and DeepSpeed Ulysses
+(see Figure 1). The next section highlights the key innovations of FPDT, 
+and subsequent sections provide additional details on the design and 
+usability of FPDT, followed by experimental results.
+
+## Key Innovations
+
+### 1. Fully Pipelined Distributed Transformer (FPDT)
+
+The core innovation of our work is the Fully Pipelined Distributed
+Transformer (FPDT). This approach leverages a pipelined sequence
+chunking, which allows for the training of LLMs with sequence lengths up
+to 2 million tokens on just 4 A100-40GB GPUs. By breaking down the
+sequence into manageable chunks and processing them in a pipelined
+manner, FPDT significantly reduces the memory footprint while
+maintaining high computational efficiency. This method ensures that the
+GPUs are utilized effectively, even when dealing with extremely long
+sequences.
+
+### 2. Memory Optimization
+
+One of the critical aspects of our approach is the comprehensive
+analysis and optimization of the memory footprint during LLM training.
+We target the reduction of redundant intermediate buffers in both the
+forward and backward passes of the training process. By optimizing the
+use of GPU and host CPU memory, we can train larger models with longer
+sequences without running into GPU memory limitations. This optimization
+is crucial for enabling the training of ultra-long context models on a
+limited number of GPUs. It is worth noting that FPDT memory optimization
+is orthogonal and complementary to model- parameter-focused memory
```

Comment:
It would be useful to say why this is orthogonal with something like FPDT optimizes memory footprint of sequence/activations while ZeRO/FSDP optimizes memory footprint of model parameters.